File: src/main/java/cn/pprocket/items/Comment.java
package cn.pprocket.items;

import cn.pprocket.HeyClient;
import cn.pprocket.utils.ParamsBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import lombok.Data;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Data
public class Comment {
    private String content;
    private String userId;
    private String commentId;
    private String userName;
    private String userAvatar;
    private boolean hasMore;
    private String replyName;
    private String replyId;
    private String createdAt;
    private List<String> images;
    private int likes;
    private String postId;
    private List<Comment> subComments;

    public List<Comment> fillSubComments() {
        if (!hasMore) return subComments;
        Map<String,String> params = new HashMap<>();
        params.put("root_comment_id",commentId);
        params.put("lastval",subComments.get(subComments.size() - 1).getCommentId());
        String url = "https://api.xiaoheihe.cn/bbs/app/comment/sub/comments?" + new ParamsBuilder(params).build("/bbs/app/comment/sub/comments/");
        String string = HeyClient.INSTANCE.get(url);
        JsonArray array = JsonParser.parseString(string).getAsJsonObject().getAsJsonArray("comments");
        array.forEach(jsonElement -> {
            Comment comment = HeyClient.INSTANCE.parseComment(jsonElement.getAsJsonObject());
            this.subComments.add(comment);
        });
        hasMore = JsonParser.parseString(string).getAsJsonObject().get("has_more").getAsBoolean();
        removeDuplicate(this.subComments);
        return this.subComments;

    }
    private static void removeDuplicate(List<Comment> comments) {
        Map<String, Comment> distinctCommentsMap = new LinkedHashMap<>();
        for (Comment comment : comments) {
            distinctCommentsMap.putIfAbsent(comment.commentId, comment);
        }
        comments.clear();
        comments.addAll(distinctCommentsMap.values());
    }
}

File: src/commonMain/kotlin/cn/pprocket/items/Comment.kt
package cn.pprocket.items

import cn.pprocket.HeyClient.domain
import cn.pprocket.HeyClient.get
import cn.pprocket.HeyClient.parseComment
import cn.pprocket.utils.ParamsBuilder
import kotlinx.serialization.json.*

class Comment {
    var content: String = ""
    var userId: String = ""
    var commentId: String = ""
    var userName: String = ""
    var userAvatar: String = ""
    var hasMore = false
    var replyName: String = "null"
    var replyId: String = ""
    var createdAt: String = ""
    var images: List<String> = emptyList()
    var likes = 0
    var postId: String = ""
    var subComments: MutableList<Comment> = mutableListOf()
    var extraPost: Post = Post()
    var liked = false
    var postTitle: String = ""

    suspend fun fillSubComments(): List<Comment>? {
        if (!hasMore) return subComments
        val params = mapOf(
            "root_comment_id" to commentId!!,
            "lastval" to subComments!![subComments!!.size - 1].commentId!!
        )
        val url =
            "${domain}/bbs/app/comment/sub/comments?" + ParamsBuilder(params).build("/bbs/app/comment/sub/comments/")
        val string = get(url)
        val array = Json.decodeFromString<JsonObject>(string)["comments"]!!.jsonArray
        array.forEach {
            val comment = parseComment(it.jsonObject)
            subComments!!.add(comment)
        }
        hasMore = Json.decodeFromString<JsonObject>(string)["has_more"]!!.jsonPrimitive.boolean
        this.subComments.distinctBy { it.commentId }
        subComments!!.forEach { ele: Comment -> ele.postId = postId }

        return this.subComments
    }

}
